lib/internal/modules/run_main.js

```diff
@@ -166,26 +167,31 @@ function executeUserEntryPoint(main = process.argv[1]) {
   if (!useESMLoader) {
     const cjsLoader = require('internal/modules/cjs/loader');
     const { Module } = cjsLoader;
-    if (getOptionValue('--experimental-detect-module')) {
-      try {
-        // Module._load is the monkey-patchable CJS module loader.
-        Module._load(main, null, true);
-      } catch (error) {
-        const source = cjsLoader.entryPointSource;
+
+    tryCatchRethrow(() => {
+      // Module._load is the monkey-patchable CJS module loader.
```

Comment:
Calling back from C++ to JS would have a non-trivial performance impact. I think the proper way to deal with this without all the weird error handling should be handling ESM directly in the CJS loader once when the compilation fails and the source looks like ESM (and if it's async, simply ignoring the exports of the entry point should be fine, because no one really needs that, or if they do, we can warn/error).

Reply:
I have a branch [here](https://github.com/joyeecheung/node/tree/esm-detect-cpp) based on https://github.com/nodejs/node/pull/52413 that implements the idea above. This moves the reparsing into C++ and also handles ESM entrypoints in-place in the CJS loader, so it passes `./node test/parallel/test-debugger-exceptions.js` when the flag is flipped because the error doesn't bubble all the way up.